### PR TITLE
Add basic support for `std::tie` and tuples in the fwd mode

### PR DESCRIPTION
**Pull request by gojakuch**

### CODE REVIEW

**Comment by gojakuch** [comment]
@vgvassilev we should deprecate c++11 as you've suggested

**Comment by vgvassilev** [comment]
> @vgvassilev we should deprecate c++11 as you've suggested

Because of using `index_sequence`?

**Comment by gojakuch** [comment]
> > @vgvassilev we should deprecate c++11 as you've suggested
> 
> Because of using `index_sequence`?

yes and also the `constexpr` stuff. I'm guessing you might suggest using some reimplementation of the `index_sequence`?

**Comment by vgvassilev** [comment]
> > > @vgvassilev we should deprecate c++11 as you've suggested
> > 
> > 
> > Because of using `index_sequence`?
> 
> yes and also the `constexpr` stuff. I'm guessing you might suggest using some reimplementation of the `index_sequence`?

Yeah, maybe in a separate compatibility header file...

File: test/ForwardMode/STLCustomDerivatives.C [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// RUN: %cladclang %s -std=c++14 -I%S/../../include -oSTLCustomDerivatives.out | %filecheck %s
+// RUN: %cladclang -std=c++17 %s -I%S/../../include -oSTLCustomDerivatives.out | %filecheck %s
```
Comment (vgvassilev) [inline test/ForwardMode/STLCustomDerivatives.C:1]: Why do we need c++17 here?

**Review by vgvassilev — COMMENTED** [review]
Can you squash all commits into a commit with an appropriate commit message?

File: test/ForwardMode/STLCustomDerivatives.C (hunk 1 above)
Comment (gojakuch) [reply test/ForwardMode/STLCustomDerivatives.C:1]: we don't. I turned it on and forgot to switch it back to c++14.

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Pull request merged**